go/app/main.go

```diff
@@ -84,54 +79,78 @@ func addItem(c echo.Context) error {
 	if _, err := io.Copy(dst, src); err != nil {
 		return c.JSON(http.StatusInternalServerError, Response{Message: err.Error()})
 	}
-	file, err = os.OpenFile("items.json", os.O_RDWR|os.O_CREATE, 0644)
-	if err != nil {
-		return c.JSON(http.StatusInternalServerError, Response{Message: err.Error()})
-	}
-	defer file.Close()
 
-	//5.jsonファイルをdecode jsonのNewDecoderとDecode関数を使う
-	decoder := json.NewDecoder(file)
-	if err := decoder.Decode(&itemlist); err != nil {
-		return c.JSON(http.StatusInternalServerError, Response{Message: err.Error()})
-	}
-	item = Item{Name: item.Name, Category: item.Category, Image: imageFilename}
+	//item = Item{Name: item.Name, Category: item.Category, Image: imageFilename}
 
 	// 6. step5でdecodeしたitemをstep3のitemに追加する
-	itemlist.Items = append(itemlist.Items, item)
+	//itemlist.Items = append(itemlist.Items, item)
 
-	// 7. jsonの書き込み用にファイルを開く
-	//fileに再代入なので:=ではなく=で書く
-	file, err = os.Create("items.json")
+	//データベースへの接続
+	db, err := sql.Open("sqlite3", DB_PATH)
 	if err != nil {
 		return c.JSON(http.StatusInternalServerError, Response{Message: err.Error()})
 	}
-	defer file.Close()
-	//8. jsonファイルに書き込み
-	encoder := json.NewEncoder(file)
-	if err := encoder.Encode(itemlist); err != nil {
+	defer db.Close()
+
+	//商品の追加
+	var categoryid int
+	row := db.QueryRow("SELECT id FROM categories WHERE name = $1", item.Category)
+	err = row.Scan(&categoryid)
+	if err != nil {
+		if err == sql.ErrNoRows {
+			_, err = db.Exec("INSERT INTO categories (name) VALUES ($1)", item.Category)
+			if err != nil {
+				return err
```

Comment:
エラーを特定しやすくするためにメッセージを追加してみましょう

```suggestion
				return .JSON(http.StatusInternalServerError, Response{Message: "failed to insert category" + err.Error()})
```

Reply:
113行目を
```go
return c.JSON(http.StatusInternalServerError, Response{Message: "failed to insert category" + err.Error()})
``` 
に変更したところ、
```go
curl -X POST \
  --url 'http://localhost:9000/items' \
  -F 'name=apple' \
  -F 'category=food' \
  -F 'image=@/Users/hamajimariho/Downloads/local_image.jpg'
{"message":"failed to insert categoryno such table: categories"}
``` 

ログ
```go
"time":"2024-02-29T10:55:59.602638+09:00","id":"","remote_ip":"127.0.0.1","host":"localhost:9000","method":"POST","uri":"/items","user_agent":"curl/7.79.1","status":500,"error":"","latency":1944792,"latency_human":"1.944792ms","bytes_in":28455,"bytes_out":65}
``` 

messageが{"message":"Internal Server Error"}→{"message":"failed to insert categoryno such table: categories"}に変わりました

97,98行目あたりが違うのかなと思いましたが、よくわかりませんでした・・・。

Reply:
`no such table: categories`

Databaseのなかに categories tableが無いのではないでしょうか？ 確認してみてください！

Reply:
```go
 sqlite3 mercari.sqlite3
SQLite version 3.37.0 2021-12-09 01:34:53
Enter ".help" for usage hints.
sqlite> .tables
categories  items     
``` 
データベースファイルを作成してスキーマを保存するところがよくわからなくて、そこがうまくいっていないのかもしれないです。